go/vt/vtgate/planbuilder/operators/info_schema_planning.go

```diff
@@ -0,0 +1,314 @@
+	cmp, ok := in.(*sqlparser.ComparisonExpr)
+	if !ok || cmp.Operator != sqlparser.EqualOp {
+		return false, "", nil
+	}
+
+	isSchemaName, col := isTableOrSchemaRoutable(cmp)
+	rhs := cmp.Right
+	if col == nil || !shouldRewrite(rhs) {
+		return false, "", nil
+	}
+
+	// here we are just checking if this query can be translated to an evalengine expression
+	// we'll need to do this translation again later when building the engine.Route
+	_, err := evalengine.Translate(rhs, &notImplementedSchemaInfoConverter{})
+	if err != nil {
+		// if we can't translate this to an evalengine expression,
+		// we are not going to be able to route based on this expression,
+		// and might as well move on
+		return false, "", nil
+	}
+	var name string
+	if isSchemaName {
+		name = sqltypes.BvSchemaName
+	} else {
+		name = reservedVars.ReserveColName(col)
+	}
+	cmp.Right = sqlparser.NewArgument(name)
+	return isSchemaName, name, rhs
+}
+
+// isTableOrSchemaRoutable searches for a comparison where one side is a table or schema name column.
+// if it finds the correct column name being used,
+// it also makes sure that the LHS of the comparison contains the column, and the RHS the value sought after
+func isTableOrSchemaRoutable(cmp *sqlparser.ComparisonExpr) (
+	isSchema bool, // tells if we are dealing with a table or a schema name comparator
+	col *sqlparser.ColName, // which is the colName we are comparing against
+) {
+	if col, schema, table := IsTableSchemaOrName(cmp.Left); schema || table {
+		return schema, col
+	}
+	if col, schema, table := IsTableSchemaOrName(cmp.Right); schema || table {
+		// to make the rest of the code easier, we shuffle these around so the ColName is always on the LHS
+		cmp.Right, cmp.Left = cmp.Left, cmp.Right
+		return schema, col
+	}
+
+	return false, nil
+}
+
+func tryMergeInfoSchemaRoutings(routingA, routingB Routing, m merger, lhsRoute, rhsRoute *Route) (ops.Operator, error) {
+	// we have already checked type earlier, so this should always be safe
+	isrA := routingA.(*InfoSchemaRouting)
+	isrB := routingB.(*InfoSchemaRouting)
+	emptyA := len(isrA.SysTableTableName) == 0 && len(isrA.SysTableTableSchema) == 0
+	emptyB := len(isrB.SysTableTableName) == 0 && len(isrB.SysTableTableSchema) == 0
+
+	switch {
+	// if either side has no predicates to help us route, we can merge them
+	case emptyA:
+		return m.merge(lhsRoute, rhsRoute, isrB)
+	case emptyB:
+		return m.merge(lhsRoute, rhsRoute, isrA)
+
+	// if both sides have the same schema predicate, we can safely merge them
+	case sqlparser.Equals.Exprs(isrA.SysTableTableSchema, isrB.SysTableTableSchema):
+		for k, expr := range isrB.SysTableTableName {
+			if e, found := isrA.SysTableTableName[k]; found && !sqlparser.Equals.Expr(expr, e) {
+				// schema names are the same, but we have contradicting table names, so we give up
+				return nil, nil
+			}
+			isrA.SysTableTableName[k] = expr
+		}
+		return m.merge(lhsRoute, rhsRoute, isrA)
```

Comment:
isn't same schema name enough to ensure it will go to same keyspace? I think we do not need to check if table schema is same.
If they are not present in the predicate then we should check for table name

Reply:
yeah, that makes sense